rust/rbac-registration/src/cardano/cip509/mod.rs

```diff
@@ -25,28 +25,43 @@ use validation::{
 use x509_chunks::X509Chunks;
 
 use super::transaction::witness::TxWitness;
-use crate::utils::{
-    decode_helper::{decode_bytes, decode_helper, decode_map_len},
-    general::{decode_utf8, decremented_index},
-    hashing::{blake2b_128, blake2b_256},
+use crate::{
+    cardano::cip509::{rbac::Cip509RbacMetadata, types::ValidationSignature},
+    utils::{
+        decode_helper::{decode_bytes, decode_helper, decode_map_len},
+        general::decremented_index,
+        hashing::{blake2b_128, blake2b_256},
+    },
 };
 
 /// CIP509 label.
 pub const LABEL: u64 = 509;
 
-/// CIP509.
-#[derive(Debug, PartialEq, Clone, Default)]
+/// A x509 metadata envelope.
+///
+/// The envelope is required to prevent replayability attacks. See [this document] for
+/// more details.
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/blob/x509-envelope-metadata/CIP-XXXX/README.md
+#[derive(Debug, PartialEq, Clone)]
 pub struct Cip509 {
-    /// `UUIDv4` Purpose .
-    pub purpose: Uuid, // (bytes .size 16)
+    /// A registration purpose (`UUIDv4`).
+    ///
+    /// The purpose is defined by the consuming dApp.
+    pub purpose: Uuid,
```

Comment:
I would rather that none of the field are public, and we expose methods which use them where required.

Reply:
Usually I also prefer this approach, but in this case we move values out of `Cip509` fields when constructing the registration chain. With private fields we will be forced to clone them. Probably not a big dial, but still fills unnecessary.